lib/Dialect/FIRRTL/LowerTypes.cpp

```diff
@@ -0,0 +1,485 @@
+                              results);
+
+  // In the base case we record this field.
+  auto bundle = type.dyn_cast<BundleType>();
+  if (!bundle) {
+    results.push_back({type, suffixSoFar.str(), isFlipped});
+    return;
+  }
+
+  SmallString<16> tmpSuffix(suffixSoFar);
+
+  // Otherwise, we have a bundle type.  Break it down.
+  for (auto &elt : bundle.getElements()) {
+    // Construct the suffix to pass down.
+    tmpSuffix.resize(suffixSoFar.size());
+    tmpSuffix.push_back('_');
+    tmpSuffix.append(elt.first.strref());
+    // Recursively process subelements.
+    flattenBundleTypes(elt.second, tmpSuffix, isFlipped, results);
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// Pass Infrastructure
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct FIRRTLTypesLowering
+    : public LowerFIRRTLTypesBase<FIRRTLTypesLowering>,
+      public FIRRTLVisitor<FIRRTLTypesLowering, LogicalResult> {
+
+  using ValueField = std::pair<Value, std::string>;
+  using FIRRTLVisitor<FIRRTLTypesLowering, LogicalResult>::visitDecl;
+  using FIRRTLVisitor<FIRRTLTypesLowering, LogicalResult>::visitExpr;
+  using FIRRTLVisitor<FIRRTLTypesLowering, LogicalResult>::visitStmt;
+
+  void runOnOperation() override;
+
+  // Lowering module block arguments.
+  void lowerArg(BlockArgument arg, BundleType type);
+  void lowerArg(BlockArgument arg, Type type);
+
+  // Lowering operations.
+  LogicalResult visitDecl(InstanceOp op);
+  LogicalResult visitExpr(SubfieldOp op);
+  LogicalResult visitStmt(ConnectOp op);
+
+  // Skip over any other operations.
+  LogicalResult visitUnhandledOp(Operation *op) { return success(); }
+  LogicalResult visitInvalidOp(Operation *op) { return success(); }
+
+  // Helpers to manage state.
+  Value addArgument(Type type, unsigned oldArgNumber,
+                    StringRef nameSuffix = "");
+
+  void setBundleLowering(Value oldValue, StringRef flatField, Value newValue);
+  Optional<Value> getBundleLowering(Value oldValue, StringRef flatField);
+  void getAllBundleLowerings(Value oldValue, SmallVectorImpl<Value> &results);
+
+  void setInstanceLowering(Value oldValue, Value newValue);
+  Value getInstanceLowering(Value oldValue);
+
+  void setSubfieldLowering(Value subfield, Value rootBundle,
+                           std::string suffix);
+  Optional<ValueField> getSubfieldLowering(Value subfield);
+
+  void removeArg(unsigned argNumber);
+  void removeOp(Operation *op);
+  void cleanup();
+
+private:
+  // The builder is set and maintained in the main loop.
+  OpBuilder *builder;
+
+  // State to keep track of arguments and operations to clean up at the end.
+  SmallVector<unsigned, 8> argsToRemove;
+  SmallVector<Operation *, 16> opsToRemove;
+
+  // State to keep a mapping from original bundle values to flattened names and
+  // flattened values.
+  circt::DenseMap<Value, llvm::StringMap<Value>> loweredBundleValues;
+
```

Comment:
Relatedly, would it be better for the key of `loweredBundleValues` to be a std::pair containing both oldValue and flatField?  It would simplify the lookup and reduce memory use vs maps of maps.  `StringMap` is pretty heavy on a per-instance basis (it's not super efficient to hold just a few strings like this).  One concern about that is that you have to make sure the strings are uniqued somewhere.  You could use a `llvm::StringSet` to do that. 

If you don't feel comfortable with the data structure gymnastics, I can do this after you land the patch (just remind me :-)

Reply:
... ah, I see that this makes `getAllBundleLowerings` more complicated.  I think that the connect lowering stuff could work in a different way to not need this.  Curious to know what you think.

Reply:
I've toyed with a few approaches here, including using a pair as the key instead of using a map of maps. And agreed, the connect lowering could be re-worked to simplify this. Let me try to simplify this and update the PR.

I'm interested in getting the data structure gymnastics correct if you don't mind reviewing while I learn :)

Reply:
I'm very happy to help you grow and learn!